File: practices/chapter9_sort/python/ch_9_merge_sort.py
def merge(left_arr=[], right_arr=[]):
  res = []

  while left_arr and right_arr:
    if left_arr[0] <= right_arr[0]:
      res.append(left_arr.pop(0))
    else:
      res.append(right_arr.pop(0))
  if left_arr:
    res += left_arr
  if right_arr:
    res += right_arr
  
  print('Merged: ', res)
  return res

def merge_sort(data_list):
  if len(data_list) <= 1:
    return data_list

  middle_index = len(data_list) // 2
  left_sort = data_list[:middle_index]
  right_sort = data_list[middle_index:]

  # 每次都切一半去sort，最後再把結果合起來
  print('Left: ', left_sort, 'Right: ', right_sort)
  left_sort = merge_sort(left_sort)
  right_sort = merge_sort(right_sort)

  return merge(left_sort, right_sort)

def main():
  data_list = [9, 3, 2, 8, 7, 10, 6, 5, 4, 1,]
  print('Origin data list: ', data_list)
  print('Merge sorted: ', merge_sort(data_list))

if __name__ == "__main__":
    main()
  

File: practices/chapter9_sort/js/ch_9_merge_sort.js
const {
  consoleSort,
} = require('./consoleSort')

function mergeTwoArr(arr1=[], arr2=[]) {
  let res = []

  while(arr1.length > 0 && arr2.length > 0) {
    if(arr1[0] <= arr2[0]) {
      res.push(arr1.shift())
    } else {
      res.push(arr2.shift())
    }
  }

  // 如果有單一arr比較多的情況
  if(arr1.length > 0) {
    res = [...res, ...arr1]
  }
  if(arr2.length > 0) {
    res = [...res, ...arr2]
  }
  // console.log(res)

  return res
}

function mergeSort(dataList=[]) {
  if(dataList.length < 2) {
    return dataList
  }
  const middleIndex = Math.floor(dataList.length / 2)
  const leftArr = dataList.slice(0, middleIndex)
  const rightArr = dataList.slice(middleIndex)
  // console.log(leftArr, rightArr)

  const leftSorted = mergeSort(leftArr)
  const rightSorted = mergeSort(rightArr)

  const res = mergeTwoArr(leftSorted, rightSorted)
  return res
}

const main = consoleSort(mergeSort)

// main()

module.exports = {
  mergeSort,
}